src/main/java/com/google/devtools/build/lib/remote/RemoteModule.java

```diff
@@ -511,6 +513,17 @@ public void afterAnalysis(
           filesToDownload.addAll(getRunfiles(configuredTarget));
         }
       }
+      // Also fetch all inputs for actions that propagate their inputs (symlink actions).
+      for (ActionInput actionInput : filesToDownload.build()) {
+        if (!(actionInput instanceof Artifact)) {
+          continue;
+        }
+        ActionExecutionMetadata action =
+            (ActionExecutionMetadata) analysisResult.getActionGraph().getGeneratingAction((Artifact) actionInput);
+        if (action.mayInsensitivelyPropagateInputs()) {
+          filesToDownload.addAll(action.getInputs().toList());
```

Comment:
Is it possible for there to be more than one symlink-ish action chained together? My concern is that this may be insufficient if the inputs to this action are outputs of other symlink actions. Rewinding recurses in such a case. Perhaps that would be overkill here.

Reply:
Good catch.